shared/src/main/scala/mlscript/helpers.scala

```diff
@@ -78,10 +127,180 @@ abstract class TypeImpl extends Located { self: Type =>
   }
   
+  /** Converts a term to its typescript type declarations, including
+    * any new type aliases created in order to represent it's type
+    *
+    * @param termName name of term that's been declared. If none is given
+    *                 default "res" is used to indicate type of value of result
+    *                 from evaluating the term
+    * @return
+    */
+  def toTsTypeSourceCode(termName: Option[String] = None): SourceCode = {
+    val ctx = ToTsTypegenContext.empty;
+    
+    // Create a mapping from type var to their friendly name for lookup
+    // Also add them as type parameters to the current type because typescript
+    // uses parametric polymorphism
+    ctx.existingTypeVars = ShowCtx.mk(this :: Nil, "").vs;
+    val tsType = this.toTsType(ctx);
+    
+    ctx.newTypeParams = ctx.newTypeParams ++ ctx.existingTypeVars.map(tup => SourceCode(tup._2));
+    val completeTypegenStatement = termName match {
+      // term definitions bound to names are exported
+      // as declared variables with their derived types
+      case S(name) =>
+        SourceCode("// start ts") +
+        SourceCode.concat(ctx.newTypeAlias) +
+        (SourceCode(s"export declare const $name") ++
+          SourceCode.paramList(ctx.newTypeParams) ++
+          SourceCode.colon ++
+          tsType
+        ) +
+        SourceCode("// end ts")
+      // terms definitions not bound to names are not exported by default
+      // they are bound to an implicit res variable and the type of res
+      // is shown here
+      case N =>
+        SourceCode("// start ts") +
+        SourceCode.concat(ctx.newTypeAlias) +
+        (SourceCode(s"type res") ++
+          SourceCode.paramList(ctx.newTypeParams) ++
+          SourceCode.equalSign ++
+          tsType
+        ) +
+        SourceCode("// end ts")
+    }
+      
+    completeTypegenStatement
+  }
+
+  /**
+    * Converts an mlscript type to source code representation of a ts type. It also
+    * keep tracks of extra type variables and type parameters defined through the context.
+    *
+    * Takes a context to maintain state.
+    * @param ctx
+    * @return
+    */
+  def toTsType(implicit ctx: ToTsTypegenContext = ToTsTypegenContext.empty): SourceCode = {
+    this match {
+      // these types can be inferred from expression that do not need to be
+      // assigned to a variable
+      case Union(TypeName("true"), TypeName("false")) | Union(TypeName("false"), TypeName("true")) => SourceCode("boolean")
+      case Union(lhs, rhs) => SourceCode.sepBy(List(lhs.toTsType, rhs.toTsType), SourceCode.separator)
+      case Inter(lhs, rhs) => SourceCode.sepBy(List(lhs.toTsType, rhs.toTsType), SourceCode.ampersand)
+      case Record(fields) => SourceCode.recordWithEntries(fields.map(entry => (SourceCode(entry._1.name), entry._2.toTsType)))
+      // unwrap extra parenthesis when tuple has single element
+      case Tuple(fields) => if (fields.length === 1) {
```

Comment:
This does not seem right. `[T]` is not the same as `T`.

Reply:
Each function type should have a tuple type as its parameter type, which corresponds to a TS parameter list. It's never okay to turn tuples of one field into their element type. You do see that this cannot be the right thing to do, right?

Reply:
To be clearer: think of the function type `(int,) -> (int,)`. It should NOT be converted to `(arg0: int) -> int`, but rather to `(arg0: int) -> [int]`. In other words, you should handle function parameter types specially.

Reply:
Also, try `def f (x, y) = x + y`. You have to get into the mindset of thinking to handle all possible situations that could arise, instead of making an over-specialized fix for one concrete example you're currently looking at.

Reply:
In that sense each argument is actually a parameter list much like Scala. Typegen now does shows the correct behaviour.